overlord/ifacestate/apparmorprompting/prompting.go

```diff
@@ -0,0 +1,578 @@
+	currentListener := m.listener
+	m.lock.Unlock()
+
+	m.tomb.Go(func() error {
+		logger.Debugf("starting listener")
+		if err := listenerRun(currentListener); err != listener.ErrClosed {
+			return err
+		}
+		return nil
+	})
+
+run_loop:
+	for {
+		logger.Debugf("waiting prompt loop")
+		select {
+		case req, ok := <-listenerReqs(currentListener):
+			if !ok {
+				// Reqs() closed, so either errored or Stop() was called.
+				// In either case, the listener Close() method has already
+				// been called, and the tomb error will be set to the return
+				// value of the Run() call from the previous tracked goroutine.
+				logger.Debugf("listener closed requests channel")
+				break run_loop
+			}
+
+			// XXX: this debug log leaks information about internal activity
+			logger.Debugf("Got from kernel req chan: %v", req)
+			if err := m.handleListenerReq(req); err != nil {
+				// XXX: this log leaks information about internal activity
+				logger.Noticef("Error while handling request: %+v", err)
+			}
+		case <-m.tomb.Dying():
+			logger.Noticef("InterfacesRequestsManager tomb is dying, disconnecting")
+			break run_loop
+		}
+	}
+	return m.disconnect()
+}
+
+func (m *InterfacesRequestsManager) handleListenerReq(req *listener.Request) error {
+	userID := uint32(req.SubjectUID)
+	// TODO: immediately deny if req.SubjectUID() == 0 (root)
+	snap := req.Label // Default to apparmor label, in case process is not a snap
+	tag, err := naming.ParseSecurityTag(req.Label)
+	if err == nil {
+		// the triggering process is not a snap, so treat apparmor label as snap field
+		snap = tag.InstanceName()
+	}
+
+	// TODO: when we support interfaces beyond "home", do a proper selection here
+	iface := "home"
+
+	path := req.Path
+
+	permissions, err := prompting.AbstractPermissionsFromAppArmorPermissions(iface, req.Permission)
+	if err != nil {
+		// XXX: this log leaks information about internal activity
+		logger.Noticef("error while parsing AppArmor permissions: %v", err)
+		response := listener.Response{
+			Allow:      false,
+			Permission: req.Permission,
+		}
+		return requestReply(req, &response)
+	}
+
+	remainingPerms := make([]string, 0, len(permissions))
+	satisfiedPerms := make([]string, 0, len(permissions))
+
+	// we're done with early checks, serious business starts now, and we can
+	// take the lock
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	matchedDenyRule := false
+	for _, perm := range permissions {
+		if yesNo, err := m.rules.IsPathAllowed(userID, snap, iface, path, perm); err == nil {
+			if !yesNo {
+				matchedDenyRule = true
+			} else {
+				satisfiedPerms = append(satisfiedPerms, perm)
+			}
+		} else {
+			if !errors.Is(err, requestrules.ErrNoMatchingRule) {
+				logger.Noticef("error while checking request against existing rules: %v", err)
+			}
+			// No matching rule found
+			remainingPerms = append(remainingPerms, perm)
+		}
+	}
+	if matchedDenyRule {
+		// XXX: this debug log leaks information about internal activity
```

Comment:
can you explain the what is the issue here XXX? isn't that the nature of debug logs?

Reply:
The concern is about logging the private application usage activity of users in the logs. But this is a debug message, so it's probably okay.

Reply:
I've removed all these comments and the actual request details from the messages, now we'll just get the errors.